File: 21. Merge Two Sorted Lists/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sort(self, list: ListNode, comp: ListNode):
        if not list.next:
            list.next = comp
            return
        else:
            if list.next.val <= comp.val:
                self.sort(list.next, comp)
            else:
                newComp = list.next
                list.next = comp
                self.sort(list.next, newComp)
        
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1 and not list2:
            return None
        elif not list1:
            return list2
        elif not list2:
            return list1
        if list1.val <= list2.val:
            out = list1
            self.sort(list1, list2)
        else:
            out = list2
            self.sort(list2, list1)
            
        return out

File: 21. Merge Two Sorted Lists/solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
 var mergeTwoLists = function (list1, list2) {
    if (!list1 && !list2) {
        return list1;
    }
    let out = new ListNode();
    let curOut = out;
    let cur1 = list1;
    let cur2 = list2;
    while (cur1 || cur2) {
        if (!cur1 && !cur2) break;
        if (cur1 && cur2) {
            if (cur1?.val <= cur2?.val) {
                curOut.val = cur1.val;
                cur1 = cur1.next;
            } else {
                curOut.val = cur2.val;
                cur2 = cur2.next;
            }
        } else if (cur1) {
            curOut.val = cur1.val;
            cur1 = cur1.next;
        } else if (cur2) {
            curOut.val = cur2.val;
            cur2 = cur2.next;
        }
        if (cur1 || cur2) {
            curOut.next = new ListNode();
            curOut = curOut.next;
        }
    }
    return out;
};